zerver/lib/user_notification_info.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import dataclass
+from typing import List, Optional, Set
+
+from zerver.models import UserProfile, query_for_ids
+
+
+@dataclass
+class UserNotificationInfo:
+    id: int
+    enable_online_push_notifications: bool
+    enable_offline_push_notifications: bool
+    enable_offline_email_notifications: bool
+    is_bot: bool
+    bot_type: Optional[int]
+    long_term_idle: bool
+
+
+class UserNotificationInfoBackend:
+    def get_user_notification_info(self, user_ids: Set[int]) -> List[UserNotificationInfo]:
```

Comment:
Could we call these something more specific, like "`UserPmMentionInfo`"? Or maybe even "`CachableUserNotificationData`" or something? I think it would be nice to somehow indicate that these objects do exclude a large part of the user's notification settings, particularly those related to streams (which are queried earlier in `get_recipient_info`).

Reply:
Naming is a bit tricky here.  I do want to keep things somewhat concise.  I don't think `UserPmMentionInfo` is better here, since this info is used for non-mention-related things`.  I think you can quibble about `Info` vs. `Data`, but this info is used to help produce the result for `get_recipient_info`, so that seems appropriate.  I don't think we need `Cachable` in the name of the type, since every data type is inherently cache-able.  And then the `Backend` suffix here makes the class parallel to `MentionBackend`.

It's hard to come up with perfect names, but these names are unique in the codebase, and they communicate that we are tracking info for users related to notifications, which I think is pretty much what they are.

Reply:
Just to be clear, since Abhijeet didn't mention it explicitly, the problem here is that this has a very similar name and somewhat similar function to the existing `UserMessageNotificationsData` class, and we need to figure out a way to make it not confusing.

Reply:
I think a good name for this data structure would be something more like `RawUserNotificationData`.  `raw` is helpful for indicating it's values direct from the database, without processing; and the rest references that it's from the User/UserProfile table.

Reply:
Similar names for similar functions sounds about right to me. Do I need to make the name less similar or more similar?

Reply:
> ... this has a very similar name and somewhat similar function to the existing ...

> Similar names for similar functions sounds about right to me. Do I need to make the name less similar or more similar?

`UserMessageNotificationsData` serves a different purpose, but stores (slightly) processed data. `RawUserNotificationData` sounds good to me :+1:

Reply:
I think critically, the `RawUserNotificationData` is global user properties, and thus not message-specific, so it's appropriate that it not have `Message` in its name, in contrast with the other class, which is.